File: src/main/java/eu/siacs/conversations/xml/XmlReader.java
package eu.siacs.conversations.xml;

import android.util.Log;
import android.util.Xml;
import com.google.common.base.Strings;
import com.google.common.collect.ImmutableMap;
import com.google.common.io.Closeables;
import eu.siacs.conversations.Config;
import im.conversations.android.xmpp.ExtensionFactory;
import im.conversations.android.xmpp.model.StreamElement;
import java.io.Closeable;
import java.io.EOFException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import org.jspecify.annotations.NonNull;
import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

public class XmlReader implements Closeable {

    private static final int XML_ELEMENT_MAX_DEPTH = 128;

    private final XmlPullParser parser;
    private InputStream is;

    public XmlReader() {
        this.parser = Xml.newPullParser();
        try {
            this.parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, true);
        } catch (XmlPullParserException e) {
            Log.d(Config.LOGTAG, "error setting namespace feature on parser");
        }
    }

    public void setInputStream(final InputStream inputStream) throws IOException {
        if (inputStream == null) {
            throw new IOException();
        }
        this.is = inputStream;
        try {
            parser.setInput(new InputStreamReader(this.is));
        } catch (XmlPullParserException e) {
            throw new IOException("error resetting parser");
        }
    }

    public void reset() throws IOException {
        if (this.is == null) {
            throw new IOException();
        }
        try {
            parser.setInput(new InputStreamReader(this.is));
        } catch (XmlPullParserException e) {
            throw new IOException("error resetting parser");
        }
    }

    @Override
    public void close() {
        final var current = this.is;
        Closeables.closeQuietly(current);
        this.is = null;
    }

    public @NonNull Tag readTag() throws IOException {
        try {
            while (this.is != null && parser.next() != XmlPullParser.END_DOCUMENT) {
                if (parser.getEventType() == XmlPullParser.START_TAG) {
                    final var id = new ExtensionFactory.Id(parser.getName(), parser.getNamespace());
                    final var attrBuilder = new ImmutableMap.Builder<String, String>();
                    for (int i = 0; i < parser.getAttributeCount(); ++i) {
                        // TODO we would also look at parser.getAttributeNamespace()
                        final var value = parser.getAttributeValue(i);
                        final var prefix = parser.getAttributePrefix(i);
                        final String name;
                        if (Strings.isNullOrEmpty(prefix)) {
                            name = parser.getAttributeName(i);
                        } else {
                            name = prefix + ":" + parser.getAttributeName(i);
                        }
                        if (name != null && value != null) {
                            attrBuilder.put(name, value);
                        }
                    }
                    return new Tag.Start(id, attrBuilder.buildKeepingLast());
                } else if (parser.getEventType() == XmlPullParser.END_TAG) {
                    final var id = new ExtensionFactory.Id(parser.getName(), parser.getNamespace());
                    return new Tag.End(id);
                } else if (parser.getEventType() == XmlPullParser.TEXT) {
                    return new Tag.No(parser.getText());
                }
            }

        } catch (final IOException e) {
            throw e;
        } catch (final Throwable throwable) {
            throw new IOException(
                    "xml parser mishandled "
                            + throwable.getClass().getSimpleName()
                            + "("
                            + throwable.getMessage()
                            + ")",
                    throwable);
        }
        throw new EOFException();
    }

    public <T extends StreamElement> T readElement(final Tag.Start current, final Class<T> clazz)
            throws IOException {
        final Element element = readElement(current);
        if (clazz.isInstance(element)) {
            return clazz.cast(element);
        }
        throw new IOException(
                String.format("Read unexpected {%s}%s", element.getNamespace(), element.getName()));
    }

    public Element readElement(final Tag currentTag) throws IOException {
        return readElement(currentTag, 0);
    }

    private Element readElement(final Tag currentTag, final int depth) throws IOException {
        if (depth >= XML_ELEMENT_MAX_DEPTH) {
            throw new XmlMaxDepthReachedException();
        }
        final ExtensionFactory.Id id;
        final Element element;
        if (currentTag instanceof Tag.Start start) {
            id = start.getId();
            element = ExtensionFactory.create(id);
            element.setAttributes(start.getAttributes());
        } else {
            throw new IOException("Cannot start reading element at tag other than start");
        }
        while (true) {
            final var tag = this.readTag();
            System.out.println("encountered inner tag: " + tag.getClass());
            switch (tag) {
                case Tag.Start innerStart -> {
                    final var child = this.readElement(innerStart, depth + 1);
                    element.addChild(child);
                }
                case Tag.No no -> {
                    if (element.getChildren().isEmpty()) {
                        element.setContent(no.getText());
                    }
                }
                case Tag.End end -> {
                    if (end.getId().equals(id)) {
                        return element;
                    } else {
                        throw new IOException("End tag did not match start tag");
                    }
                }
                default -> throw new IOException("Read invalid tag " + tag.getClass());
            }
        }
    }

    public static class XmlMaxDepthReachedException extends IOException {
        public XmlMaxDepthReachedException() {
            super("Reached maximum depth of XML stream");
        }
    }
}

File: src/main/java/eu/siacs/conversations/xml/XmlReader.kt
package eu.siacs.conversations.xml

import android.util.Xml
import com.google.common.base.Strings
import com.google.common.collect.ImmutableMap
import com.google.common.io.Closeables
import im.conversations.android.xmpp.ExtensionFactory
import im.conversations.android.xmpp.model.StreamElement
import java.io.Closeable
import java.io.EOFException
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException

class XmlReader : Closeable {

    private var parserInputStream: ParserInputStream? = null

    fun setInputStream(inputStream: InputStream) {
        this.parserInputStream = of(inputStream)
    }

    fun reset() {
        val current = this.parserInputStream
            ?: throw IOException("Unable to reset. No current parser")
        this.parserInputStream = of(current.inputStream)
    }

    override fun close() {
        val current = this.parserInputStream
        if (current != null) {
            this.parserInputStream = null
            Closeables.closeQuietly(current.inputStream)
        }
    }

    @Throws(IOException::class)
    fun readTag(): Tag {
        try {
            while (parserInputStream != null &&
                parserInputStream!!.parser.next() != XmlPullParser.END_DOCUMENT) {
                val parser = parserInputStream!!.parser
                if (parserInputStream!!.parser.eventType == XmlPullParser.START_TAG) {
                    val id = ExtensionFactory.Id(parser.name, parserInputStream!!.parser.namespace)
                    val attrBuilder = ImmutableMap.Builder<String, String>()
                    for (i in 0 until parser.attributeCount) {
                        val value = parser.getAttributeValue(i)
                        val prefix = parser.getAttributePrefix(i)
                        val name: String =
                            if (Strings.isNullOrEmpty(prefix)) {
                                parser.getAttributeName(i)
                            } else {
                                "$prefix:${parser.getAttributeName(i)}"
                            }
                        if (value != null) {
                            attrBuilder.put(name, value)
                        }
                    }
                    return Tag.Start(id, attrBuilder.buildKeepingLast())
                } else if (parser.eventType == XmlPullParser.END_TAG) {
                    val id = ExtensionFactory.Id(parser.name, parser.namespace)
                    return Tag.End(id)
                } else if (parser.eventType == XmlPullParser.TEXT) {
                    return Tag.No(parser.text)
                }
            }
        } catch (e: IOException) {
            throw e
        } catch (throwable: Throwable) {
            throw IOException(
                "xml parser mishandled ${throwable.javaClass.simpleName}(${throwable.message})",
                throwable
            )
        }
        throw EOFException()
    }

    @Throws(IOException::class)
    fun <T : StreamElement> readElement(current: Tag.Start, clazz: Class<T>): T {
        val element = readElement(current)
        if (clazz.isInstance(element)) {
            return clazz.cast(element)
        }
        throw IOException(
            String.format("Read unexpected {%s}%s", element.namespace, element.name)
        )
    }

    @Throws(IOException::class)
    fun readElement(currentTag: Tag.Start): Element {
        return readElement(currentTag, 0)
    }

    @Throws(IOException::class)
    private fun readElement(parent: Tag.Start, depth: Int): Element {
        if (depth >= XML_ELEMENT_MAX_DEPTH) {
            throw XmlMaxDepthReachedException()
        }
        val id = parent.id
        val element = ExtensionFactory.create(id)
        element.setAttributes(parent.attributes)
        while (true) {
            when (val tag = this.readTag()) {
                is Tag.Start -> {
                    val child = this.readElement(tag, depth + 1)
                    element.addChild(child)
                }
                is Tag.No -> {
                    if (element.getChildren().isEmpty()) {
                        element.setContent(tag.text)
                    }
                }
                is Tag.End -> {
                    if (tag.id == id) {
                        return element
                    } else {
                        throw IOException("End tag did not match start tag")
                    }
                }
                else -> throw IOException("Read invalid tag ${tag.javaClass}")
            }
        }
    }

    class XmlMaxDepthReachedException : IOException("Reached maximum depth of XML stream")

    private data class ParserInputStream(val parser: XmlPullParser, val inputStream: InputStream)

    companion object {
        private const val XML_ELEMENT_MAX_DEPTH = 128

        @Throws(IOException::class)
        private fun of(inputStream: InputStream): ParserInputStream {
            val parser = Xml.newPullParser()
            try {
                parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, true)
                parser.setInput(InputStreamReader(inputStream))
                return ParserInputStream(parser, inputStream)
            } catch (e: XmlPullParserException) {
                throw IOException(e)
            }
        }
    }
}
